### PR TITLE
docs(DatePicker, TimePicker): explain constraint validation

### PR DESCRIPTION
**Pull request by vursen**
## Description

- [x] Give a general description of how the constraint validation works, including such aspects as validation triggers, constraints, error messages, Binder, manual validation.
- [x] For setErrorMessage, mention i18n as the means for setting distinct error messages.
- [x] For setErrorMessage, explain that the error message will be only displayed when the field is 
- [x] For constraint methods like setMin, setMax, mention that they are inclusive.

Part of https://github.com/vaadin/flow-components/issues/6545

## Type of change

- [x] Docs

### CODE REVIEW

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 1]
```diff
@@ -82,6 +82,32 @@
...
+ * <h3>Validation</h3>
+ * <p>
+ * DatePicker comes with a basic validation mechanism based on constraints.
+ * Validation is triggered when the user attempts to apply a date whether by
+ * entering it manually or selecting it from the overlay, or when the date is
+ * updated programmatically. Validation checks if the date is parsable and
+ * satisfies the defined constraints. If validation fails, the component is
+ * marked as invalid and an error message is displayed (if one is provided).
+ * <p>
+ * The following constraints are available:
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:94]:
```suggestion
 * The following constraints can be set:
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 2]
```diff
@@ -82,6 +82,32 @@
...
+ * marked as invalid and an error message is displayed (if one is provided).
+ * <p>
+ * The following constraints can be set:
+ * <ul>
+ * <li>{@link #setRequiredIndicatorVisible(boolean)}
+ * <li>{@link #setMin(LocalDate)}
+ * <li>{@link #setMax(LocalDate)}
+ * </ul>
+ * <p>
+ * Error messages for constraint violations, including unparsable input, can be
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:101]:
```suggestion
 * Error messages for constraint violations and unparsable input, can be
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 3]
```diff
@@ -82,6 +82,32 @@
...
+ * marked as invalid and an error message is displayed (if one is provided).
+ * <p>
+ * The following constraints can be set:
+ * <ul>
+ * <li>{@link #setRequiredIndicatorVisible(boolean)}
+ * <li>{@link #setMin(LocalDate)}
+ * <li>{@link #setMax(LocalDate)}
+ * </ul>
+ * <p>
+ * Error messages for constraint violations and unparsable input, can be
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:101]:
```suggestion
 * Error messages for constraint violations and unparsable input can be
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 4]
```diff
@@ -82,6 +82,32 @@
...
+ * marked as invalid and an error message is displayed (if one is provided).
+ * <p>
+ * The following constraints can be set:
+ * <ul>
+ * <li>{@link #setRequiredIndicatorVisible(boolean)}
+ * <li>{@link #setMin(LocalDate)}
+ * <li>{@link #setMax(LocalDate)}
+ * </ul>
+ * <p>
+ * Error messages for constraint violations and unparsable input can be
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:101]:
```suggestion
 * Error messages for unparsable input and constraint violations can be
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 5]
```diff
@@ -82,6 +82,32 @@
...
+ * marked as invalid and an error message is displayed (if one is provided).
+ * <p>
+ * The following constraints can be set:
+ * <ul>
+ * <li>{@link #setRequiredIndicatorVisible(boolean)}
+ * <li>{@link #setMin(LocalDate)}
+ * <li>{@link #setMax(LocalDate)}
+ * </ul>
+ * <p>
+ * Error messages for unparsable input and constraint violations can be provided
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:101]:
```suggestion
 * Error messages for unparsable input and constraint violations can be configured
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 6]
```diff
@@ -82,6 +82,32 @@
  * the format of the current locale or through the date picker overlay. The
  * overlay opens when the field is clicked and/or any input is entered when the
  * field is focused.
+ * <h3>Validation</h3>
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:85]:
```suggestion
 * <h2>Validation</h2>
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 7]
```diff
@@ -82,6 +82,32 @@
...
+ * <p>
+ * The following constraints can be set:
+ * <ul>
+ * <li>{@link #setRequiredIndicatorVisible(boolean)}
+ * <li>{@link #setMin(LocalDate)}
+ * <li>{@link #setMax(LocalDate)}
+ * </ul>
+ * <p>
+ * Error messages for unparsable input and constraint violations can be
+ * configured using the {@link DatePickerI18n} object.
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:102]:
```suggestion
 * configured using the respective methods in the {@link DatePickerI18n} object.
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 8]
```diff
@@ -82,6 +82,32 @@
...
+ * <li>{@link #setMin(LocalDate)}
+ * <li>{@link #setMax(LocalDate)}
+ * </ul>
+ * <p>
+ * Error messages for unparsable input and constraint violations can be
+ * configured using the respective methods in the {@link DatePickerI18n} object.
+ * <p>
+ * In addition to validation, constraints may also have a visual representation.
+ * For example, dates before the minimum date appear disabled in the overlay.
+ * <p>
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:106]: **open question:** Does Binder need to be mentioned?

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 9]
```diff
@@ -82,6 +82,32 @@
  * the format of the current locale or through the date picker overlay. The
  * overlay opens when the field is clicked and/or any input is entered when the
  * field is focused.
+ * <h2>Validation</h2>
+ * <p>
+ * DatePicker comes with a built-in validation mechanism based on constraints.
+ * Validation is triggered when the user attempts to apply a date whether by
+ * entering it manually or selecting it from the overlay, or when the date is
+ * updated programmatically. Validation checks if the date is parsable and
+ * satisfies the defined constraints. If validation fails, the component is
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:91]:
```suggestion
 * satisfies the specified constraints. If validation fails, the component is
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 10]
```diff
@@ -82,6 +82,32 @@
...
+ * <h2>Validation</h2>
+ * <p>
+ * DatePicker comes with a built-in validation mechanism based on constraints.
+ * Validation is triggered when the user attempts to apply a date whether by
+ * entering it manually or selecting it from the overlay, or when the date is
+ * updated programmatically. Validation checks if the date is parsable and
+ * satisfies the specified constraints. If validation fails, the component is
+ * marked as invalid and an error message is displayed (if one is provided).
+ * <p>
+ * The following constraints can be set:
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:94]:
```suggestion
 * The following constraints are supported:
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 11]
```diff
@@ -82,6 +82,32 @@
...
+ * <p>
+ * The following constraints are supported:
+ * <ul>
+ * <li>{@link #setRequiredIndicatorVisible(boolean)}
+ * <li>{@link #setMin(LocalDate)}
+ * <li>{@link #setMax(LocalDate)}
+ * </ul>
+ * <p>
+ * Error messages for unparsable input and constraint violations can be
+ * configured using the respective methods in the {@link DatePickerI18n} object.
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:102]:
```suggestion
 * configured using the respective properties in the {@link DatePickerI18n} object.
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 12]
```diff
@@ -82,6 +82,32 @@
  * the format of the current locale or through the date picker overlay. The
  * overlay opens when the field is clicked and/or any input is entered when the
  * field is focused.
+ * <h2>Validation</h2>
+ * <p>
+ * DatePicker comes with a built-in validation mechanism based on constraints.
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:87]:
```suggestion
 * Date Picker comes with a built-in validation mechanism based on constraints.
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 13]
```diff
@@ -82,6 +82,32 @@
...
+ * <p>
+ * The following constraints are supported:
+ * <ul>
+ * <li>{@link #setRequiredIndicatorVisible(boolean)}
+ * <li>{@link #setMin(LocalDate)}
+ * <li>{@link #setMax(LocalDate)}
+ * </ul>
+ * <p>
+ * Error messages for unparsable input and constraint violations can be
+ * configured using the respective properties in the {@link DatePickerI18n} object.
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:102]:
```suggestion
 * configured with the respective properties in the {@link DatePickerI18n} object. If you want to
 * provide a single error message for all constraints, you can also use the
 * {@link #setErrorMessage(String)} method. Note that error messages set with
 * {@link #setErrorMessage(String)} will take priority over i18n error messages.
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 14]
```diff
@@ -82,6 +82,35 @@
...
+ * The following constraints are supported:
+ * <ul>
+ * <li>{@link #setRequiredIndicatorVisible(boolean)}
+ * <li>{@link #setMin(LocalDate)}
+ * <li>{@link #setMax(LocalDate)}
+ * </ul>
+ * <p>
+ * Error messages for unparsable input and constraint violations can be
+ * configured with the respective properties in the {@link DatePickerI18n} object. If you want to
+ * provide a single error message for all constraints, you can also use the
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:103]:
```suggestion
 * provide a single error message for all constraints, including unparsable input, you can also use the
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 15]
```diff
@@ -82,6 +83,43 @@
...
+ * {@link #setErrorMessage(String)} will take priority over i18n error messages
+ * if both are set.
+ * <p>
+ * In addition to validation, constraints may also have a visual representation.
+ * For example, dates before the minimum date appear disabled in the overlay.
+ * <p>
+ * For more complex validation that requires custom rules, you can use
+ * {@link Binder}. By default, before running custom validators, Binder will
+ * also validate if the date is parsable and satisfies the component constraints
+ * using error messages from the component's i18n object.
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:116]:
```suggestion
 * For more advanced validation that requires custom rules, you can use
 * {@link Binder}. By default, before running custom validators, Binder will
 * also check if the date is parsable and satisfies the component constraints
 * using error messages from the component's i18n object.
```

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 16]
```diff
@@ -339,13 +376,29 @@ public DatePicker(LocalDate initialDate, Locale locale) {
         setLocale(locale);
     }
 
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Distinct error messages depending on constraints and for unparsable input
```
Comment (vursen) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:382]:
```suggestion
     * Distinct error messages for unparsable input and different constraints
```

**Review by rolfsmeds — CHANGES_REQUESTED** [review]
See my comments in the corresponding TextField PR :)

**Comment by vursen** [comment]
> See my comments in the corresponding TextField PR :)

Updated DatePicker based on the TextField PR.

**Review by web-padawan — APPROVED** [review]

**Review by web-padawan — APPROVED** [review]

**Review by web-padawan — APPROVED** [review]

**Review by rolfsmeds — APPROVED** [review]

**Pull request merged**